File: binary_search/js/binarySearch.js
function identity(x) { return x; }

function findIndex(elements, value, key = identity) {
  let left = 0, right = elements.length - 1;
  while (left <= right) {
    const middle = Math.floor((left + right) / 2);
    const middleElement = key(elements[middle]);
    if (middleElement === value) return middle;
    if (middleElement < value) left = middle + 1;
    else right = middle - 1;
  }
  return null;
}

function contains(elements, value, key = identity) {
  return findIndex(elements, value, key) !== null;
}

function find(elements, value, key = identity) {
  const idx = findIndex(elements, value, key);
  return idx === null ? null : elements[idx];
}

function findLeftmostIndex(elements, value, key = identity) {
  let index = findIndex(elements, value, key);
  if (index !== null) {
    while (index >= 0 && key(elements[index]) === value) index--;
    index++;
  }
  return index;
}

function findRightmostIndex(elements, value, key = identity) {
  let index = findIndex(elements, value, key);
  if (index !== null) {
    while (index < elements.length && key(elements[index]) === value) index++;
    index--;
  }
  return index;
}

function findAllIndices(elements, value, key = identity) {
  const left = findLeftmostIndex(elements, value, key);
  const right = findRightmostIndex(elements, value, key);
  if (left === null || right === null) return new Set();
  return new Set(Array.from({ length: right - left + 1 }, (_, i) => left + i));
}

function findLeftmost(elements, value, key = identity) {
  const idx = findLeftmostIndex(elements, value, key);
  return idx === null ? null : elements[idx];
}

function findRightmost(elements, value, key = identity) {
  const idx = findRightmostIndex(elements, value, key);
  return idx === null ? null : elements[idx];
}

function findAll(elements, value, key = identity) {
  const indices = findAllIndices(elements, value, key);
  return new Set(Array.from(indices).map(i => elements[i]));
}

module.exports = {
  identity,
  findIndex,
  contains,
  find,
  findLeftmostIndex,
  findRightmostIndex,
  findAllIndices,
  findLeftmost,
  findRightmost,
  findAll,
};

File: binary_search/python/binary_search.py
from operator import attrgetter
from typing import Any, Callable, Iterable, Optional, Set

# Utilidad básica
def identity(x):
    return x

# Función principal de búsqueda
def find_index(elements: Iterable[Any], value: Any, key: Callable = identity) -> Optional[int]:
    left, right = 0, len(elements) - 1
    while left <= right:
        middle = (left + right) // 2
        middle_element = key(elements[middle])
        if middle_element == value:
            return middle
        elif middle_element < value:
            left = middle + 1
        else:
            right = middle - 1
    return None

# Funciones auxiliares
def contains(elements, value, key: Callable = identity) -> bool:
    return find_index(elements, value, key) is not None

def find(elements, value, key: Callable = identity):
    idx = find_index(elements, value, key)
    return None if idx is None else elements[idx]

def find_leftmost_index(elements, value, key: Callable = identity) -> Optional[int]:
    index = find_index(elements, value, key)
    if index is not None:
        while index >= 0 and key(elements[index]) == value:
            index -= 1
        index += 1
    return index

def find_leftmost(elements, value, key: Callable = identity):
    idx = find_leftmost_index(elements, value, key)
    return None if idx is None else elements[idx]

def find_rightmost(elements, value, key: Callable = identity):
    idx = find_rightmost_index(elements, value, key)
    return None if idx is None else elements[idx]

def find_all_indices(elements, value, key: Callable = identity) -> Set[int]:
    left = find_leftmost_index(elements, value, key)
    right = find_rightmost_index(elements, value, key)
    if left is None or right is None:
        return set()
    return set(range(left, right + 1))

def find_all(elements, value, key=identity):
    return [elements[i] for i in find_all_indices(elements, value, key)]

def find_rightmost_index(elements, value, key: Callable = identity) -> Optional[int]:
    index = find_index(elements, value, key)
    if index is not None:
        while index < len(elements) and key(elements[index]) == value:
            index += 1
        index -= 1
    return index